### ISSUE TITLE
GraphQL documentation syntax not supported

### ISSUE THREAD

**Issue by robross0606** [AEB-labs/cruddl#96]
cruddl appears to be throwing errors on Graphql documentation syntax.  If my schema has something like this:
```
"""
IS08601 time duration
"""
scalar Duration
```
Processing through cruddl reports errors like this:
```
[error]: uncaughtException: Project has errors:
Error: This kind of definition is not allowed. Only object and enum type definitions are allowed. at schema.graphqls:2:1
```

**Comment by robross0606** [comment]
FYI, I also tried with alternate syntax like this:
```
"IS08601 time duration"
scalar Duration
```
However, the same error occurs.

**Comment by robross0606** [comment]
Actually, now I think it is just the use of `scalar` at all that is causing the issue.

**Comment by Yogu** [comment]
Yes, custom scalars are not supported by cruddl. To implement it, we first need to decide how they can be configured by the schema author, because we can't just use callbacks for validation (cruddl never uses callbacks for modelling). We could e.g. allow to specify a regular expression to validate scalars. What do you think

**Comment by robross0606** [comment]
Something like regex would be a start so we could at least support `scalar` definitions to some degree.  However, the lack of callbacks is problematic for several reasons.  Validation is one.  But there are others such as audit logging that could become really problematic for integration of cruddl.  

I know you don't support it, so my thought was to integrate some of those things (i.e. audit logging) as other custom directives so I could chain things together.  cruddl is a great idea, but it can't be an impenetrable line between graphql and arango.  Developers need to be able to inject and extract data along that pipeline.

**Comment by robross0606** [comment]
Hmmm... Now I'm not even sure cruddl will allow the addition of other directive handlers.

**Comment by robross0606** [comment]
Okay, so this is a non-starter as well.  Cruddl throws up if you also reference other non-cruddl directives.  By graphql standard, @directives are supposed to be cumulative and chainable from left-to-right.  There is no reason cruddl should be throwing an error if it sees a directive it doesn't understand.  Is there a way to pass other directive handlers into cruddl so it knows about the use of other non-cruddl directives such as [graphql-constraint-directive](https://github.com/confuser/graphql-constraint-directive).

**Comment by Yogu** [comment]
> Something like regex would be a start so we could at least support scalar definitions to some degree. However, the lack of callbacks is problematic for several reasons. Validation is one. But there are others such as audit logging that could become really problematic for integration of cruddl.

I didn't write cruddl does not use callbacks for configuration - it already does in several instances (just look at `ProjectConfig`, it's full of optional callbacks). One callback is `profileConsumer` which gets called on each query - might go into the direction of audit logging.

I meant that *modelling* should not require callbacks. Maybe the goals of cruddl and your requirements are not exactly overlapping here - we use cruddl in a generic component called "Model Manager" where users can dynamically define their schema via a web interface. They can't just implement a cruddl callback.

However, I'm ok with callbacks and extensive configuration if it's not directly coupled to the cruddl type definitions in the `ProjectSource`s. For example, we could make the system scalars configurable. That would be pretty simple actually - it's just an array of `GraphQLScalarType` classes. We could easily make this list configurable.

>  it can't be an impenetrable line between graphql and arango. Developers need to be able to inject and extract data along that pipeline.

This line is the opposite of impemetrable! Actually, that's the part of cruddl that's closes to a "plugin" architecture. When creating a schema, you specify a database adapter. There, you can pass custom database adapters. For example, you can create a subclass of `ArangoDBAdapter` (or use composition) and hook yourself into `executeExt`. This method gets passed something called `QueryTree` which is an abstract representation of the query or mutation to be executed. You can e.g. modify the query tree or do audit checks, and then pass it to the original ArangoDB Adapter.

However, the modelling features currently don't follow a plugin architecture. In an earlier version of cruddl, we used a pipeline approach to gradually "enhance" the cruddl-input to the real, executable GraphQL schema. However, that approach got pretty messy pretty quickly, and we developed a framework that allows us to implement features in a cleaner way. 

> Hmmm... Now I'm not even sure cruddl will allow the addition of other directive handlers.

If we manage to sketch an elegant solution to really support custom directives, this can make its way to cruddl. But it first needs to be designed properly. For example, we could allow users to specify custom directive definitions, which then are made available in the model classes. Then, you could read these directives within a custom database adapter. However, if you want to dive deep into the core cruddl concepts of root entities, child entities, value objects, entity extensions, references, relations, collect fields etc. and modify their behavior, this would be very complicated to design properly I think.

> By graphql standard, @Directives are supposed to be cumulative and chainable from left-to-right. There is no reason cruddl should be throwing an error if it sees a directive it doesn't understand.

Not sure what part of the GraphQL spec you are referring to. We use the validator `KnownDirectives` from `graphql-js` with the following documentation:

> A GraphQL document is only valid if all `@directives` are known by the schema and legally positioned.

Also, the `sources` cruddl accepts use GraphQL syntax and adhere to its semantics, but they are by design not exactly equivalent to arbitrary type system definitions. We're just using GraphQL to configure cruddl. A very important goal of cruddl is that schema errors are caught with meaningful messages, assisting the user in writing functional schemas. That's why everything that's not explicitly implemented is forbidden. Just allowing to specify directive definitions would not help at all - the directives would just get lost in cruddl.

**Comment by robross0606** [comment]
The pipeline approach is really how directives are supposed to work, at least in Apollo.  If you look at how directive handlers are generally implemented, they follow an approach where one directive can handle a piece of data and then pass it off to the next directive, ultimately ending in the resolver (if any).  It is a chain of handlers.  The problem right now is that cruddl appears to provide no way to add known directive types other than what cruddl already knows.  It cuts off the ability to leverage other directive components such as [graphql-constraint-directive](https://github.com/confuser/graphql-constraint-directive) which would be a natural fit to validate data BEFORE cruddl even gets it to pass to Arango.  I don't think you need to necessarily support a plugin architecture since the architecture to register handlers is already provided by Apollo.  What would be better is to allow cruddl to play nice in that architecture.  

Even if cruddl simply ignored directives it didn't understand instead of throwing an error, that would be better.  That would at least allow us to have cruddl create the executable schema it uses and pass that off to something like [merge-graphql-schemas](https://github.com/Urigo/merge-graphql-schemas) for integration with other directives or even other parts of the overall schema that cruddl isn't intended to handle.

It is the heavy-handed validation during `project.createSchema()` that is causing the problem since cruddl precludes the possibility that there may be other parts of the code handling some of the directives.  Apollo supports registering as many directive handlers as you want, but cruddl doesn't allow it.
